packages/target-tauri/translationfn/src/lib.rs

```diff
@@ -0,0 +1,278 @@
+    }
+
+    #[test]
+    fn three_substitutions_scrambled() {
+        let msgs: HashMap<String, HashMap<String, String>> =
+            serde_json::from_value(serde_json::json!({"group_name_changed_by_other": {
+              "message": "%3$s已将群组名称从“%1$s”更改为“%2$s”"
+            }}))
+            .unwrap();
+        let txen = TranslationEngine::new(msgs, "zh_CN").unwrap();
+        let result = txen.translate(
+            "group_name_changed_by_other",
+            Substitution::String(vec!["1", "2", "3"]),
+        );
+        assert_eq!(result, "3已将群组名称从“1”更改为“2”");
+    }
+
+    #[test]
+    fn four_substitutions() {
+        let msgs: HashMap<String, HashMap<String, String>> =
+            serde_json::from_value(serde_json::json!({"aeap_addr_changed": {
+              "message": "%1$s hat die E-Mail-Adresse von %2$s nach %3$s geändert und %4$s"
+            }}))
+            .unwrap();
+        let txen = TranslationEngine::new(msgs, "de").unwrap();
+        let result = txen.translate(
+            "aeap_addr_changed",
+            Substitution::String(vec![
+                "Max Mustermann",
+                "max@mustermann.de",
+                "mustermann@max.de",
+                "ist toll",
+            ]),
+        );
+        assert_eq!(
+            result,
+            "Max Mustermann hat die E-Mail-Adresse von max@mustermann.de nach mustermann@max.de geändert und ist toll"
+        );
+    }
+    #[test]
+    fn quantity_ru() {
+        let msgs: HashMap<String, HashMap<String, String>> =
+            serde_json::from_value(serde_json::json!({ "n_minutes": {
+              "one": "%d минуту",
+              "few": "%d минуты",
+              "many": "%d минут",
+              "other": "%d минут"
+            }}))
+            .unwrap();
+        let txen = TranslationEngine::new(msgs, "ru").unwrap();
+        let result = txen.translate("n_minutes", Substitution::Quantity(1));
+        assert_eq!(result, "1 минуту");
+
+        let result = txen.translate("n_minutes", Substitution::Quantity(2));
+        assert_eq!(result, "2 минуты");
+
+        let result = txen.translate("n_minutes", Substitution::Quantity(5));
+        assert_eq!(result, "5 минут");
+
+        let result = txen.translate("n_minutes", Substitution::QuantityFloat(1.5));
+        assert_eq!(result, "1.5 минут");
+    }
+    #[test]
+    fn quantity_en() {
+        let msgs: HashMap<String, HashMap<String, String>> =
+            serde_json::from_value(serde_json::json!({ "n_hours": {
+              "one": "%d hour",
+              "other": "%d hours"
+            }}))
+            .unwrap();
+        let txen = TranslationEngine::new(msgs, "en").unwrap();
+        let result = txen.translate("n_hours", Substitution::Quantity(1));
+        assert_eq!(result, "1 hour");
+
+        let result = txen.translate("n_hours", Substitution::Quantity(2));
+        assert_eq!(result, "2 hours");
+    }
+
+    //TODO: test for dead errors to not crash and are returned
```

Comment:
This comment is not clear to me, but maybe it's just me.

Reply:
```suggestion
    // TODO: test for that errors to not crash and are returned as errors instead (creation of translation engine)
    // TODO: test that errors of translate function do not crash and just return the key (not sure if we can check if logging worked, might be too difficult/complex, so just checking for not panicking is enough)
```